jmclient/jmclient/blockchaininterface.py

```diff
@@ -285,18 +285,34 @@ def import_addresses_if_needed(self, addresses, wallet_name):
             self.import_addresses(addresses - imported_addresses, wallet_name)
         return import_needed
 
-    def _yield_transactions(self, wallet_name):
-        batch_size = 1000
-        iteration = 0
+    def _yield_transactions(self):
+        """ Generates a lazily fetched sequence of transactions seen in the
+        wallet (under any label/account), yielded in newest-first order. Care
+        is taken to avoid yielding duplicates even when new transactions are
+        actively being added to the wallet while the iteration is ongoing.
+        """
+        num, skip = 1, 0
+        txs = self.list_transactions(num, skip)
+        if not txs:
+            return
+        tx = txs[0]
+        yield tx
         while True:
-            new = self._rpc(
-                'listtransactions',
-                ["*", batch_size, iteration * batch_size, True])
-            for tx in new:
+            num *= 2
+            txs = self.list_transactions(num, skip)
+            if not txs:
```

Comment:
I want to check my understanding of the logic here. Isn't it impossible to reach this? On line 295 we call list_transactions with (1, 0) meaning we pull the most recent tx if there is one (and if there isn't, we already returned on line 297). Here, we call with (2, 0) (skip has not changed, right), so there's no way we can end up with a null response here, right? Even if a new transaction came in the mean time.

Reply:
> skip has not changed, right

On the first time through the loop that is true.

> there's no way we can end up with a null response here, right?

We can if transactions disappeared from the Bitcoin Core wallet between our RPCs. It's there to guard against an infinite loop. Consider what would happen if the entry we're looking for (to re-sync with what we've already yielded) were to disappear from the Bitcoin Core wallet entirely. The loop will continue fetching entries all the way back to the beginning, not finding the last entry it had yielded. At some point it will call `listtransactions` with `skip` >= the number of entries in the wallet, at which point the RPC will return no results. If we didn't check for that, then we'd just keep calling the RPC over and over trying to get more results to find the last entry we had yielded.